pkg/service/model_build_listener.go

```diff
@@ -0,0 +1,137 @@
+package service
+
+import (
+	"context"
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/aws-load-balancer-controller/pkg/annotations"
+	elbv2model "sigs.k8s.io/aws-load-balancer-controller/pkg/model/elbv2"
+	"strconv"
+)
+
+func (t *defaultModelBuildTask) buildListeners(ctx context.Context) error {
+	cfg := t.buildListenerConfig(ctx)
+	for _, port := range t.service.Spec.Ports {
+		_, err := t.buildListener(ctx, port, cfg)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (t *defaultModelBuildTask) buildListener(ctx context.Context, port corev1.ServicePort, cfg listenerConfig) (*elbv2model.Listener, error) {
+	lsSpec, err := t.buildListenerSpec(ctx, port, cfg)
+	if err != nil {
+		return nil, err
+	}
+	listenerResID := fmt.Sprintf("%v", port.Port)
+	ls := elbv2model.NewListener(t.stack, listenerResID, lsSpec)
+	return ls, nil
+}
+
+func (t *defaultModelBuildTask) buildListenerSpec(ctx context.Context, port corev1.ServicePort, cfg listenerConfig) (elbv2model.ListenerSpec, error) {
+	tgProtocol := elbv2model.Protocol(port.Protocol)
+	listenerProtocol := elbv2model.Protocol(port.Protocol)
+	if tgProtocol != elbv2model.ProtocolUDP && cfg.certificateARNs != nil && (cfg.tlsPortsSet.Len() == 0 ||
```

Comment:
for empty check for lists, we should always use `len(cfg. certificateARNs) != 0`, which handles both cases.